packages/@aws-cdk/pipelines/lib/actions/publish-assets-action.ts

```diff
@@ -157,6 +157,10 @@ export class PublishAssetsAction extends CoreConstruct implements codepipeline.I
         CDK_CLI_VERSION: { value: props.cdkCliVersion },
       } : undefined,
     });
+
+    App.addCustomSynthesis(this, {
```

Comment:
Why not just use a `Stage.of(this).outdir` and push the build spec file to the cloud assembly in the constructor itself? It feels like we can avoid implementing this in `synthesize()`?

Reply:
A totally legitimate approach (used by asset bundling for example)

Reply:
BuildSpec won't be known yet at constructor time.

Reply:
We could use `Lazy`.